ex2_hal/iris/hardware_interface/source/iris.c

```diff
@@ -0,0 +1,460 @@
+                ret = get_data(image_length_buffer, MAX_IMAGE_LENGTH);
+                if (ret == IRIS_HAL_OK) {
+                    controller_state = FINISH;
+                } else {
+                    controller_state = ERROR_STATE;
+                }
+                /* It is expected that the first byte in the buffer will be the LSB */
+                *(image_length) = (uint32_t)((uint8_t)image_length_buffer[2]<<16 | (uint8_t)image_length_buffer[1]<<8 | (uint8_t)image_length_buffer[0]); // Concatenate image_length_buffer
+
+                free(image_length_buffer);
+                controller_state = FINISH;
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on get image length command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on get image length command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a transfer image command to Iris, and expects to receive the
+ *   image data of size image_length
+ *
+ * @param[in] image_length
+ *   Number of bytes expected to be received from Iris
+ *
+ * @return
+ *   Returns IRIS_HAL_OK if equipment handler returns IRIS_ACK, else IRIS_HAL_ERROR
+ **/
+IrisHALReturn iris_transfer_image(uint32_t image_length) {
+    uint32_t num_transfer;
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND: // Send start image transfer command
+            {
+                ret = send_command(IRIS_TRANSFER_IMAGE);
+                if (ret == IRIS_ACK) {
+                    controller_state = GET_DATA;
+                } else {
+                    controller_state = FINISH;
+                }
+                break;
+            }
+            case GET_DATA: // Get image data in chunks/blocks
+            {
+                uint16_t * image_data_buffer = (uint16_t*) pvPortMalloc(IMAGE_TRANSFER_SIZE * sizeof(uint16_t));
+                memset(image_data_buffer, 0, IMAGE_TRANSFER_SIZE);
+                if (image_data_buffer == NULL) {
+                    ex2_log("Failed attempt to dynamically allocate memory under iris get image data");
+                    return IRIS_HAL_ERROR;;
+                }
+                num_transfer = (IMAGE_TRANSFER_SIZE + image_length) / IMAGE_TRANSFER_SIZE; // Ceiling division
+                for (uint32_t count_transfer = 0; count_transfer < num_transfer; count_transfer++) {
+                    ret = get_data(image_data_buffer, IMAGE_TRANSFER_SIZE);
+                    // TODO: Do something with the received data (e.g transfer it to the SD card)
+                    // Or just get the data and send it forward to the next stage. Prefer not to have too
```

Comment:
Would be better to add a void *buf parameter to this function, and copy the data in to that

Reply:
Sorry for the function are talking about?

Reply:
Will be implemented once service is written